src/libsyntax/diagnostic.rs

```diff
@@ -573,32 +548,27 @@ impl EmitterWriter {
 
         // If we elided something, put an ellipsis.
         if display_lines < all_lines {
-            let last_line_index = display_line_infos.last().unwrap().line_index;
             let s = format!("{}:{} ", fm.name, last_line_index + 1);
             try!(write!(&mut self.dst, "{0:1$}...\n", "", s.len()));
         }
 
-        // FIXME (#3260)
-        // If there's one line at fault we can easily point to the problem
-        if lines.lines.len() == 1 {
-            let lo = cm.lookup_char_pos(sp.lo);
-            let mut digits = 0;
-            let mut num = (lines.lines[0].line_index + 1) / 10;
-
-            // how many digits must be indent past?
-            while num > 0 { num /= 10; digits += 1; }
-
+        // FIXME (#8706)
+        if let Some(orig) = fm.get_line(last_line_index) {
             let mut s = String::new();
             // Skip is the number of characters we need to skip because they are
             // part of the 'filename:line ' part of the previous line.
-            let skip = fm.name.chars().count() + digits + 3;
-            for _ in 0..skip {
-                s.push(' ');
-            }
-            if let Some(orig) = fm.get_line(lines.lines[0].line_index) {
-                let mut col = skip;
-                let mut lastc = ' ';
-                let mut iter = orig.chars().enumerate();
+            let skip = fm.name.chars().count() + digits + 2;
+            s.extend(::std::iter::repeat(' ').take(skip));
+            let mut col = skip;
+            let mut lastc = ' ';
+            let mut iter = orig.chars().enumerate().peekable();
+
+            for sp in msp.spans.as_slice().iter() {
+                let lo = cm.lookup_char_pos(sp.lo);
+                if lo.line != (last_line_index + 1) {
+                    // Only render spans starting on the last line
```

Comment:
I guess this is fine for now since we currently group non-line-overflowing spans into single lines anyway.